### PR TITLE
feat: macos signing and notarization 

### PR DESCRIPTION
**Pull request by lidel**
> This is a streamlined  refactor of #362 on road to fix issues described in  #8240

This PR adds `sign-macos` job to CI workflow, adding a transparent signing step. 

The main reason is to sign repo migrations and avoid issue from https://github.com/ipfs/go-ipfs/issues/8240 globally (outside of Brave). It also provides us with a template if we wish to add signing for other platforms.

### Technical details

This PR splits `make publish` performed in `main.yml` into three stages:
1. `build` executes  `./dockerized make all_dists` and produces any missing binaries in `./releases`
2. `sign-macos` signs and notarizes every new binary from `./releases` and updates `.tar.gz` packages and relevant hashes in-place
3. `publish` 
   * executes `make publish` and adds any new and  signed packages from `./releases` to `DIST_ROOT` 
     (`/ipns/dist.ipfs.io` by default)
   * pins produced CID to ipfs-cluster for persistence

### Flow for adding new dist or version

1. Open a PR against this repo with new entry for one of `./dists/{name}/versions` 
2. Wait for CI to build signed  dist, put it on IPFS and pin updated `DIST_ROOT` to ipfs-cluster
3. Preview link will be published on PR  and printed at the end of `publish` step of CI build
4. If everything looks good, write down the CID  from preview link, and update the DNSlink of dist.ipfs.io

### Content routing optimization

IPFS node running on CI manually pre-connects to ipfs-websites cluster peers, which are known to already have all blocks required during dist.ipfs.io build. This way we skip DHT lookup.

### Website preview on PRs

Every PR now gets a link to website preview at the `dweb.link` IPFS gateway:

> 

## TODO

- [x] sign darwin binaries and put them in the same place as unsigned ones
- [x] ensure all hashes and manifests match signed binaries
- [x] ensure the root CID for updated dist.ipfs.io website is persisted after CI job ends
  - we use `/dnsaddr/ipfs-websites.collab.ipfscluster.io` but [it errors right now](https://github.com/ipfs/distributions/runs/3128450306?check_suite_focus=true#step:14:13)
- [x] gateway link with preview of updated website
- [x] remove fake `DIST_ROOT` and ensure CI is no-op when no new binaries are added
- [x] squash & merge

### CODE REVIEW

File: .github/workflows/main.yml
Comment (lidel) [inline .github/workflows/main.yml:14]: This is so we run build against  CID that represents dist.ipfs.io without ipfs-update@v1.7.1 – this demonstrates how additional entry in `dists/ipfs-update/versions`  triggers build of "missing" binary artifacts.

(we will remove this before merging this PR)

File: .github/workflows/main.yml
Comment (BigLep) [inline .github/workflows/main.yml:15]: What's the process for updating these?  Do we rely on someone manually updating it?  What's their trigger for doing so?

File: Dockerfile
Comment (BigLep) [inline Dockerfile:4]: Similarly, what's the process for updating this?
Can we have a URL like https://dist.ipfs.io/go-ipfs/latest-full-release/go-ipfs_linux-amd64.tar.gz

File: README.md
Comment (BigLep) [inline README.md:133]: Can we link to the issue that tracks this?

File: build-go.sh
Comment (BigLep) [inline build-go.sh:14]: nit: Seems odd in one script to overwrite a variable like this.  Maybe do as one line?

File: .github/workflows/main.yml
Comment (BigLep) [inline .github/workflows/main.yml:155]: Is it normal to have big shell scripts inlined like this in GitHub workflow?  Or would it be better to factor it out?

(I don't know  conventions here).

Also, do we have a "3 finger rule" around bash (i.e., once bash code is more than 3 fingers of lines on the terminal, we use a different language that is more developer friendly)? (We don't need to introduce that here - just curious - feel free to disregard()

**Review by BigLep — COMMENTED** [review]
Thanks for doing this!  I'll leave it for others to approve.

File: .github/workflows/main.yml
Comment (olizilla) [inline .github/workflows/main.yml:288]: This mechanism for adding a status result to a PRs checks list stopped working for me in ipfs-dns-deploy. I tried to fix it, but could never get the status to appear again on the PR... what did you change to make it work here!?

File: .github/workflows/main.yml
Comment (olizilla) [reply .github/workflows/main.yml:155]: I would support such a thing, but i don't believe there are any existing conventions for it.

File: .github/workflows/main.yml
Comment (lidel) [reply .github/workflows/main.yml:15]: I think those are low importance: could be kept as-is until someone cares to update or there is a breaking change that forces version bump.

Having static versions of build tools makes the builds more robust and reproducible, but we could also switch to always run on the latest version for dogfooding – don't feel strongly either way, but leaning towards more reproducibility.

File: .github/workflows/main.yml
Comment (lidel) [reply .github/workflows/main.yml:288]: @olizilla Hm.. not sure, could be one of below:
* copied it from https://github.com/ipfs-shipyard/ipfs-github-action/blob/182f36dcd28a0df011ce1ee9b004a1ca7fa7c720/entrypoint.sh#L11-L32 and tweaked quotes and renamed some fields.
* stopped installing `jq`  via snap and use the version shipped with `ubuntu-latest`. 
* use `GITHUB_TOKEN` secret directly, and not via env
  * I would guess this is most likely the cause. Github hardened some things around their tokens this year, [enabling orgs to tweak defaults or limit the scope](https://github.blog/changelog/2021-04-20-github-actions-control-permissions-for-github_token/) (maybe PL changed something?). I've read that some github actions stopped working with regular `GITHUB_TOKEN` and require creation of a [a PAT (Personal Access Token)](https://docs.github.com/en/enterprise-server@3.1/github/authenticating-to-github/keeping-your-account-and-data-secure/creating-a-personal-access-token) (I think only official one under `actions/*`  continue to work "old way", but also did not dig into this dep enough to understand what exactly changed and why)

File: README.md
Comment (lidel) [reply README.md]: Added https://github.com/ipfs/distributions/issues/372 in f115f93

File: Dockerfile
Comment (lidel) [reply Dockerfile]: Good catch. Simplified in 7d222b5 which makes sure HTTP client here uses same version as `GO_IPFS_VER` – just in case we have breaking change  in any of used APIs.

File: .github/workflows/main.yml
Comment (lidel) [reply .github/workflows/main.yml:155]: I feel shell scripts are the right tool for the job in CI workflows, but this could be my personal bias :)
Considered extracting those longer blocks to `.sh` files in  `./scripts/ci/` but it felt harder to read and reason about – left them inlined.

We can extract them after this + nightlies lands, but suggest doing that in a separate PR, to make reviews at hand easier by having everything in one place, removing need for jumps between files.

File: build-go.sh
Comment (lidel) [reply build-go.sh]: Simplified in 9acff7e :+1:

File: .github/workflows/main.yml
Comment (lidel) [reply .github/workflows/main.yml:155]: Update: after reviewing #369 I believe I need to extract those bash blocks into reusable scripts to land everything anyway :) (want to run them in both `main.yml` for release / PR builds and `nightly.yml` for daily matrix build).
Stay tunned.

Comment (lidel) [reply .github/workflows/main.yml]: Done. Extracted scripts are in `./scripts/ci` and will be shared with nightly  setup.

File: scripts/ci/github-preview-link.sh
Comment (lidel) [inline scripts/ci/github-preview-link.sh:13]: :information_source: we could replace this script with JS call and  [`actions/github-script`](https://github.com/actions/github-script) but I don't want to sink any more time into refactoring this PR ;)

File: dockerized
Comment (lidel) [inline dockerized:11]: :information_source: this controls the version of the CLI client _inside_ `./dockerized` build, which talks to go-ipfs running _outside_ on the host. Here, we ensure it uses the same version (with fallback to the latest one if `GO_IPFS_VER` is undefined).

File: scripts/ci/pin-to-cluster.sh
Comment (lidel) [inline scripts/ci/pin-to-cluster.sh:21]: :information_source: to make CI faster and avoid hanging forever if parts of cluster are out of sync, cluster pinning is considered "done" after we have 2 copies. 

Here, I do it manually in userland, but filled https://github.com/ipfs/ipfs-cluster/issues/1427 to add this sort of thing to ipfs-cluster-ctl CLI tool.

File: scripts/ci/setup-ipfs.sh
Comment (lidel) [inline scripts/ci/setup-ipfs.sh:10]: :information_source: This was pretty annoying. Entire build failed because Githubs DNS proxying cache at `127.0.0.*`  randomly timeouted while resolving `_dnslink` or `_dnsaddr`. Switching to Cloudflare here make the problem go away.

File: scripts/ci/sign-new-macos-releases.sh
Comment (lidel) [inline scripts/ci/sign-new-macos-releases.sh:30]: :information_source: I registered the wildcard `bundle_id=io.ipfs.dist.*`  at Apple, which enables us to have each package "isolated" from each other. This is important in case Apple decides to block something from running on macOS – we don't have all egs in one basket.

File: .github/workflows/main.yml
Comment (lidel) [inline .github/workflows/main.yml:72]: :information_source: :policeman:  I hardcoded specific revision because this is a third-party action and in theory someone could swap-out git tag and steal our signing keys.

File: .github/workflows/main.yml
Comment (lidel) [inline .github/workflows/main.yml:69]: :information_source: [gon](https://github.com/mitchellh/gon) package is used by Terraform project, so multiple eyes are looking at this

**Review by lidel — COMMENTED** [review]
Ready for review, aiming for merging in the beginning of the next week.

File: README.md
Comment (lidel) [inline README.md:102]: :information_source:   I believe this is the TLDR of this entire PR :)

File: README.md
Comment (aschmahmann) [inline README.md:218]: TODO (post PR): Let's add some references to tools for how to do the binary stripping so people don't have to search for it.

**Review by aschmahmann — APPROVED** [review]
LGTM, although I mostly reviewed the README. We may learn more as we start using it for nightlies though ❤️

File: dockerized
Comment (BigLep) [reply dockerized:11]: Should this comment live in the source file?

File: scripts/ci/pin-to-cluster.sh
Comment (BigLep) [reply scripts/ci/pin-to-cluster.sh:21]: Similar to above: when do we decide to put this in the PR vs. the code itself?  Do folks tend to do a git blame and look for comments to find more context?

**Review by BigLep — APPROVED** [review]
Great work!
Only comments were around inlining comments PR into the code for discoverability.  Your call.

File: dockerized
Comment (lidel) [reply dockerized:11]: I believe it is good to have it here in case someone else tried to modify those scripts in the future   – it explains why we install go-ipfs twice (daemon + client in docker)

File: scripts/ci/pin-to-cluster.sh
Comment (lidel) [reply scripts/ci/pin-to-cluster.sh:21]: I often use git blame,  but in this specific case those comments are provided mostly to make PR review easier.

File: README.md
Comment (lidel) [reply README.md:218]: Filled https://github.com/ipfs/distributions/issues/377 so we dont forget ;)

File: scripts/ci/pin-to-cluster.sh
Comment (BigLep) [reply scripts/ci/pin-to-cluster.sh:21]: Ok.  I'll defer to the team for the standard.  I was bringing this up because it was hard for me to imagine a case where a comment is useful during PR where it isn't also useful 6 months in the future when looking at the code with fresh eyes.  Anyways, sounds good.

**Comment by lidel** [comment]
As agreed during stewards call, I'm starting the merge dance starting with this PR

**Pull request merged**
